java/ql/lib/semmle/code/java/Member.qll

```diff
@@ -340,6 +340,46 @@ class Callable extends StmtParent, Member, @callable {
   }
 }
 
+/**
+ * Holds if the given type is public and not a nested type.
+ * If the given type is a nested type, holds if its enclosing type
+ * is also public.
+ */
+private predicate veryPublic(RefType t) {
+  t.isPublic() and
+  (
+    not t instanceof NestedType or
+    veryPublic(t.(NestedType).getEnclosingType())
+  )
```

Comment:
```suggestion
  forex(RefType r | r = t.getEnclosingType*() | r.isPublic)
```
Could this be changed to a `forex`? As in `everything enclosing it, including itself, is public`.

Reply:
There's definitely a potential performance difference. Constructions with `forall`/`forex` can sometimes be tricky performance-wise, as they are expanded into not-exists-not, and the resulting complex negations means additional materialisation of negations bodies. Furthermore, computing the TC is superfluous, since we only care about unary reachability. As such, the existing code that simply does unary reachability with a simple recursion is  almost guaranteed to be optimal.